### ISSUE TITLE
(WCAG 2.1) Required fields should be identified on first page load.

### ISSUE THREAD

**Issue by jeanniefu** [civiform/civiform#2629]
Feedback from a11y experts during an a11y audit.

Severity (Critical/Serious/Moderate/Good to have): Serious

Required fields should be identified on first page load. We should identify required fields both visually and programmatically. 

Visually, we could: (1) put a "required" text label, or (2) mark it with a star. Programmatically, we should mark the html as required, either with (1) html required attribute, or (2) aria-required attribute. The html attribute is preferred, because it has baked in information.

**Comment by clouser2** [comment]
I definitely agree with the motivation here. One challenge we're going to have lies with how we validate fields for optional question types.

For example, the Name question type has three fields: first name, middle name, and last name. If the top-level Name question is marked as optional, it's valid to submit with all these fields blank. However, if the "first name" field has a value, then the "last name" field is required for purposes of validation. As such, these fields would all be "not required" on page load, but could become "required" after the user types any content into the text box. If it's acceptable for the requiredness to change dynamically with content, we could likely write some JavaScript to enforce this.

I could imagine something like the relevant `*QuestionRenderer` class adding a `data-conditionally-required` attribute and the JS adding change handlers to each of the fields within a question containing this attribute and updating an `aria-required` attribute for all fields whenever any one of them has content (and correspondingly removing it when all fields are empty).

**Comment by jeanniefu** [comment]
Thanks so much Brett for the detailed comments! To double check my understanding, are you proposing a client-side change, so that e.g. the first/last name would dynamically be marked with aria-required without needing to wait for the user to click submit?

In terms of UI, @rahaghas do you have any preference on how to mark a question as required on first page load? Maybe we could start by marking all top level required questions with a star.

**Comment by rahaghas** [comment]
Yes I think a "*" at the end of each required question is very common so that would be great

**Comment by clouser2** [comment]
@jeanniefu Yah, I was imagining some sort of client-side validation to add/remove the `aria-required` attribute based on the current value of the other inputs for the question.

**Comment by jeanniefu** [comment]
We should start with adding an asterisk at the question level since this is straightforward.

Then followup with field level asterisk with light client side code as suggested by @clouser2.

**Comment by jeanniefu** [comment]
This was also flagged during an accessibility audit against WCAG 2.1. We will need to fix it to meet WCAG 2.1 guidelines.
